sdk/core/Azure.Core/src/Shared/ConnectionString.cs

```diff
@@ -0,0 +1,248 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Azure.Core
+{
+    internal sealed class ConnectionString
+    {
+        private readonly Dictionary<string, string> _segments;
+        private readonly string _segmentSeparator;
+        private readonly string _keyValueSeparator;
+
+        public static ConnectionString Parse(string connectionString, string segmentSeparator = ";", string keyValueSeparator = "=", bool allowFlags = false, bool allowEmptyValues = false, bool allowWhitespaces = false)
+        {
+            Validate(connectionString, segmentSeparator, keyValueSeparator, allowFlags, allowEmptyValues, allowWhitespaces);
+            return new ConnectionString(ParseSegments(connectionString, segmentSeparator, keyValueSeparator), segmentSeparator, keyValueSeparator);
+        }
+
+        private ConnectionString(Dictionary<string, string> pairs, string segmentSeparator, string keyValueSeparator)
+        {
+            _segments = pairs;
+            _segmentSeparator = segmentSeparator;
+            _keyValueSeparator = keyValueSeparator;
+        }
+
+        public string GetRequired(string key)
+        {
+            if (_segments.TryGetValue(key, out var value))
+            {
+                return value ?? throw new InvalidOperationException($"Required segment with the key '{key}' is declared as flag.");
+            }
+
+            throw new InvalidOperationException($"Required segment with the key '{key}' is missing in connection string.");
+        }
+
+        public bool HasFlag(string key)
+        {
+            if (!_segments.TryGetValue(key, out var value))
+            {
+                return false;
+            }
+
+            if (value != null)
+            {
+                throw new InvalidOperationException($"Required segment with the key '{key}' isn't a flag and has value '{value}'.");
```

Comment:
This error would be customer-facing, right?  If so, key and flag need to be common terminology around connection strings.  I can't find anything using these words with these meanings in a search on Azure connection strings:

https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string
https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-get-connection-string
https://docs.microsoft.com/en-us/azure/mysql/howto-connection-string

Reply:
Fixed.